src/main/java/com/powsybl/openloadflow/sa/LimitViolationManager.java

```diff
@@ -180,16 +176,25 @@ private void detectBusViolations(LfBus bus) {
         double busV = bus.getV();
         if (!Double.isNaN(bus.getHighVoltageLimit()) && busV > bus.getHighVoltageLimit()) {
             LimitViolation limitViolation1 = new LimitViolation(bus.getVoltageLevelId(), LimitViolationType.HIGH_VOLTAGE, bus.getHighVoltageLimit() * scale,
-                    (float) 1., busV * scale);
+                    (float) 1., busV * scale, createViolationLocation(bus));
             addBusLimitViolation(limitViolation1, bus);
         }
         if (!Double.isNaN(bus.getLowVoltageLimit()) && busV < bus.getLowVoltageLimit()) {
             LimitViolation limitViolation2 = new LimitViolation(bus.getVoltageLevelId(), LimitViolationType.LOW_VOLTAGE, bus.getLowVoltageLimit() * scale,
-                    (float) 1., busV * scale);
+                    (float) 1., busV * scale, createViolationLocation(bus));
             addBusLimitViolation(limitViolation2, bus);
         }
     }
 
+    public static ViolationLocation createViolationLocation(LfBus bus) {
+        List<Integer> nodes = bus.getNodes();
+        if (nodes.isEmpty()) {
+            return null;
+        }
+
+        return new NodeBreakerViolationLocation(bus.getVoltageLevelId(), nodes);
+    }
```

Comment:
In Bus Breaker topology you should return a BusBreakerViolationDetection.

Reply:
done, but i had to add a new method `getTopologyKind`  for all sub-class of LfBus

Reply:
I pushed a commit that moves getLocation in LfBus. It reduces the numbre of new methods and encapsulate all the topolog logic into the LfBus.
I also added the case of the breaker LfNetwork (happens in Security Analysis if switches need to be open or closed in contingencies or actions).
Let me know if it works for you.

Reply:
Looks good! Moving getLocation to LfBus is better.

Reply:
i just added a commit for test coverage

Reply:
I added more tests to check the values, including in OLF breaker mode.